### DISCUSSION TITLE
Switching to a Union Type causes cache.updateQuery to return data as null

### DISCUSSION THREAD

**Discussion by Undistraction** [urql-graphql/urql#3418] [Q&A]
We are using the graphcache exchange and this has been working really well for us for a while. However we've just made a change to the API which has meant moving from a list containing a single type, to a a list containing a Union type:

```
fragment ChatMessage_chatMessage on ChatMessage {
  faqs {
    nodes {
      ...Faq_faq
    }
  }
}
```

To a union:
```
fragment ChatMessage_chatMessage on ChatMessage {
  faqs {
    nodes {
      ... on Faq {
          ...Faq_faq
          __typename
        }
        ... on Resource {
          ...Resource_resource
          __typename
        }
    }
  }
}
```

We have a query that gets all the `ChatMessages` for an `Enrolment` which is run once the app has initialised (The `EnrolmentForChatMessagesQuery` referenced below). 

When we create a new `ChatMessage` with `createChatMessage`, the new message is returned and we append it to the current list of chat messages in an update function. This was working great before switching to a union type and whereas before `data` was the expected value, it is now always `null`. 
```
updates: {
    Mutation: {
      createChatMessage(result, { params: { enrolmentId } }, cache) {
        cache.updateQuery(
          {
            query: EnrolmentForChatMessagesQuery,
            variables: {
              id: enrolmentId,
            },
          },
          (data) => {
            // data is now null
            data.enrolment.chatMessages.nodes.push(chatMessage)
            return data
          }
        )
      }
    }
  }
```

Would really appreciate some help in tracking down the issue as I'm unsure how to debug it. It doesn't seem like a type-change should have this effect, but this is the only change. If I replace the revert this change, `data` comes back as expected. 

Would providing a schema to graphcache help? Our API is behind auth and we couldn't expose our API schema by including a schema in the src so we'd have to introspect at runtime which is not idea.

**Comment by Undistraction** [comment]
I added a schema and the problem is resolved. It would be good to understand exactly why that fixed the problem.

Given that each item in the list includes a `_typename` it certainly doesn't _feel_ like a schema should be necessary for this.

**Reply by JoviDeCroock** [reply]
You should have gotten a warning in the console about `heuristic fragment matching` basically our cache could not figure out how the data correlated I presume 😅
